lessons/cryptography/cryptography-hashing-checksums-salts.md

```diff
@@ -0,0 +1,82 @@
+
+
+## Password Salting
+Now that you know that there are many lists of commonly used passwords, you may wonder what happens if two people pick the same password. It is definitely possible for two users with different usernames to pick the same password. Usually, when you create an account, the service checks whether or not the username you pick is already taken, but it does not check if your chosen password is already being used by another user. Thus, if you look at a very large list of everyone's usernames and password hashes, you may notice that there are multiple usernames that are associated with the same common password hash. If this list is very, very large, it is highly likely that the most common hash corresponds to the most commonly used password. There are many published lists of commonly used passwords and their likelihood to be used in a given year. [NordPass](https://nordpass.com/most-common-passwords-list/) has a list of common passwords, and you can even sort it by country. Although there is likely not a perfect correlation between number of occurrences  of a password and how generally common a password is, it would not take an attacker very long to figure out what the top passwords are. 
+
+One way to help prevent the same hash from appearing in a stored list even when multiple users choose the same password is to use a technique called password salting. A salt is a random string of characters that is added to a user's password prior to hashing. Since the salt is always random, this would result in different password hashes even if multiple users choose the same password. Although the salt is a random string of letters, it is only generated randomly the first time when a user initially creates an account. After that, the salt information would be stored along with the username and hashed password, so that when the user signs in in the future, the salt would be added to the password, then hashed, and then compared with the stored value in the list. Adding a salt to every user's password not only helps prevent the same hashes from appearing multiple times, but also takes away the usefulness of the lists of common passwords and their hashes. 
+
+# The Real World, Prolific Breaches
+A good cryptographic hash function should never result in the same hash value from two different messages. To find weaknesses of the cryptographic hash algorithms, people attempt to find two input messages that would produce the same hash value. If this occurs, it is called a collision. The MD5 hash has been shown to have many collisions that are possible. With today's computing power, it is even possible to find collisions within seconds. About 20 years after the SHA-1 hash was introduced, a group of researches at Google and the CWI Institute in Amsterdam showed that they were able to provide two different PDF files that have different content but result in the same SHA-1 hash value. 
```

Comment:
Instead of the word good, I'd use something more discrete like "valid" or "proper." If a hashing algorithm produces the same output of multiple inputs it is no longer a valid algorithm. Ie: MD5